### PR TITLE
EOS-17642: implement an integration test for all2all connections

### PR DESCRIPTION
**Pull request by sergey-shilov**
This integration test involves Hare and Motr. The test
checks that after a specific sequence of HA messages
all processes are connected with each other.

As current implementation of Hare does not support
required protocol then some HA messages are sent by
a special script.

Test steps:
 - check that Hare is installed;
 - initialise loop (storage) devices;
 - bootstrap the cluster using Hare;
 - check m0traces that no HA messages are handled;
 - send a trigger HA message "TRANSIENT" to activate
   HA messages handling by DTM0 services;
 - check m0traces;
 - send "ONLINE" HA messages to trigger connection
   procedures;
 - check m0traces that connection procedures completed;
 - cleanup loop devices.

Signed-off-by: Sergey Shilov <sergey.shilov@seagate.com>

### CODE REVIEW

**Review by ivan-alekhin — APPROVED** [review]
It looks like a good base for our integration tests. If there is anything to polish, we can address it in consequent patches.

**Review by just-now — COMMENTED** [review]
Overall logic looks fine, still the productization level of the code is not sufficient to be merged.

File: dtm0/it/all2all/all2all [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+M0D_PIDS=()
+
+POOL_WIDTH=4
+IPOOL_WIDTH=2
+
+. ${MOTR_ROOT}/scripts/addb-py/chronometry/common/common_funcs
+
+
+function check_hare()
```
[the remark this answers is not in the document]
Comment (sergey-shilov) [reply dtm0/it/all2all/all2all:23]: To check whether Hare is installed properly and warn the user in other case.

File: dtm0/it/all2all/all2all [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    for i in $(seq $imgs_cnt) ; do
+        local img_file="$LOOP_IMG_DIR/disk$i.img"
+        dd if=/dev/zero of="$img_file" \
+           bs=1M seek=$(( 1 * 1024 - 1 )) count=1 &>/dev/null
+    done
+}
+
+function setup_loop_devices()
+{
+    local imgs_cnt=$((POOL_WIDTH + IPOOL_WIDTH))
```
[the remark this answers is not in the document]
Comment (sergey-shilov) [reply dtm0/it/all2all/all2all:87]: Done.

File: dtm0/it/all2all/all2all [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+    _info "Setting up loop devices"
+    for i in $(seq $imgs_cnt) ; do
```
[the remark this answers is not in the document]
Comment (sergey-shilov) [reply dtm0/it/all2all/all2all:90]: Done.

File: dtm0/it/all2all/all2all [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        fi
+    done
+}
+
+function cleanup_loop_images()
+{
+    local imgs_cnt=$((POOL_WIDTH + IPOOL_WIDTH))
+
+    _info "Removing file images"
+    for i in $(seq $imgs_cnt) ; do
```
[the remark this answers is not in the document]
Comment (sergey-shilov) [reply dtm0/it/all2all/all2all:122]: Done.

File: motr/st/utils/ha_msg_send.sh [hunk 5]
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
+
+# set -x
```
[the remark this answers is not in the document]
Comment (sergey-shilov) [reply motr/st/utils/ha_msg_send.sh:3]: Done.

File: motr/st/utils/ha_msg_send.sh [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    local lnet_nid=`sudo lctl list_nids | head -1`
+    local c_endpoint="$lnet_nid:$M0HAM_CLI_EP"
+    local s_endpoint="$lnet_nid:$1"
+    local fid=$2
+    local state=$3
+
+    # Generate HA event
+    echo "Send HA event for motr"
+    echo "c_endpoint is : $c_endpoint"
+    echo "s_endpoint is : $s_endpoint"
```
[the remark this answers is not in the document]
Comment (sergey-shilov) [reply motr/st/utils/ha_msg_send.sh:25]: Removed.

File: motr/st/utils/ha_msg_send.sh [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+    send_ha_events "$fid" "$state" "$s_endpoint" "$c_endpoint"
+}
+
+# {0x72| ((^r|1:14), ..., "192.168.122.122@tcp:12345:2:2", ...
+# {0x72| ((^r|1:10), ..., "192.168.122.122@tcp:12345:2:1", ...
+
+# EXAMPLE: ha_msg_send.sh "12345:2:2" "^r|1:26" "transient"
+# EXAMPLE: ha_msg_send.sh "12345:2:3" "^r|1:12" "online"
```
[the remark this answers is not in the document]
Comment (sergey-shilov) [reply motr/st/utils/ha_msg_send.sh:34]: Added a simple usage.

File: dtm0/it/all2all/test.yaml [hunk 8]
```diff
@@ -0,0 +1,37 @@
+nodes:
```
[the remark this answers is not in the document]
Comment (sergey-shilov) [reply dtm0/it/all2all/test.yaml:1]: Configuration is draft and will be changed when 0 spares is supported.

File: dtm0/it/all2all/all2all [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+    # send HA ONLINE
+    _info "Sending ONLINE notifications to trigger connections logic..."
+    ha_msg_send_online
+
+    _info "Checking traces..."
+    check_ne "DTM0 service: connected" 2
+    if [ $CHECK -ne 0 ]
+    then
+        fail "Process is not connected, exiting"
+    fi
```
[the remark this answers is not in the document]
Comment (sergey-shilov) [reply dtm0/it/all2all/all2all:229]: Done.

File: dtm0/it/all2all/all2all [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+    set_m0d_pids
+
+    _info "Checking traces..."
+    check_ne "evented_proc_state" 0
+    if [ $CHECK -ne 0 ]
+    then
+        fail "HA event handled instead of to be skipped, exiting."
+    fi
+
+    # send HA TRANSIENT
```
[the remark this answers is not in the document]
Comment (sergey-shilov) [reply dtm0/it/all2all/all2all:209]: Removed.

File: dtm0/it/all2all/all2all [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+    check_hare
+    init_loop_devices
+
+    _info "Bootstrapping the cluster using Hare..."
+    bootstrap_cluster
+
+    set_m0d_pids
+
+    _info "Checking traces..."
+    check_ne "evented_proc_state" 0
```
[the remark this answers is not in the document]
Comment (sergey-shilov) [reply dtm0/it/all2all/all2all:203]: Done.

[the remark this answers is not in the document]
Comment (sergey-shilov) [reply dtm0/it/all2all/all2all:203]: Done.

File: dtm0/it/all2all/all2all [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+
+function main()
+{
+    check_hare
+    init_loop_devices
+
+    _info "Bootstrapping the cluster using Hare..."
+    bootstrap_cluster
+
+    set_m0d_pids
```
[the remark this answers is not in the document]
Comment (sergey-shilov) [reply dtm0/it/all2all/all2all:200]: Done.

File: dtm0/it/all2all/all2all [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+        m0d_dir=${M0D_DIR_COMMON}${endp}
+        pid=`ls -l ${m0d_dir}/m0trace.* | awk '{ print $NF; }' | awk -F"." '{ print $2; }' | awk  'BEGIN { pid=0; } { if (pid < $0) pid = $0; } END { print pid; }'`
+        M0D_PIDS+=($pid)
+        pids+="$pid "
+    done
+
+    _info "m0d PIDs: $pids"
+}
+
+function ha_msg_send_transient()
```
[the remark this answers is not in the document]
Comment (sergey-shilov) [reply dtm0/it/all2all/all2all:145]: Added a function that gets endpoints/fids from conf.xc.

File: dtm0/it/all2all/all2all [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+MOTR_ROOT=$(realpath ../../..)
+MOTR_ST_UTILS_DIR=${MOTR_ROOT}/motr/st/utils/
+MOTR_VAR_DIR=/var/motr
+TEST_ROOT=$MOTR_VAR_DIR/all2all_test
+CURRENT_CDF=$PWD/test.yaml
+CONFD_XC=/var/lib/hare/confd.xc
+LOOP_IMG_DIR=$TEST_ROOT
+
+M0D_DIR_COMMON=$MOTR_VAR_DIR/m0d-0x7200000000000001:0x
+M0D_IDS=(c 1a 28)
```
Comment (just-now) [inline dtm0/it/all2all/all2all:14]: extract from hctl status

File: dtm0/it/all2all/all2all [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+    #systemctl reset-failed hare-hax
+    set -e
+}
+
+function bootstrap_cluster()
+{
+    hctl bootstrap --mkfs $CURRENT_CDF
+}
+
+#============= DEVICES SECTION =============#
```
Comment (just-now) [inline dtm0/it/all2all/all2all:57]: this section is not needed,
call from m0setup needed function and do not copy/paste them

File: dtm0/it/all2all/all2all [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+{
+    _info "Removing file images"
+    for i in $(seq $IMGS_CNT) ; do
+        local img_file="$LOOP_IMG_DIR/disk$i.img"
+        rm -f "$img_file"
+    done
+}
+############################################
+
+function get_m0d_pids()
```
Comment (just-now) [inline dtm0/it/all2all/all2all:125]: use pgrep alike aproach

File: dtm0/it/all2all/all2all [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+        m0d_dir=${M0D_DIR_COMMON}${endp}
+        pid=`ls -l ${m0d_dir}/m0trace.* | awk '{ print $NF; }' | awk -F"." '{ print $2; }' | awk  'BEGIN { pid=0; } { if (pid < $0) pid = $0; } END { print pid; }'`
+        M0D_PIDS+=($pid)
+        pids+="$pid "
+    done
+
+    _info "m0d PIDs: $pids"
+}
+
+function get_params_for_ha_msgs()
```
Comment (just-now) [inline dtm0/it/all2all/all2all:141]: use hctl status to generate this data.

File: dtm0/it/all2all/all2all [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+
+    for i in `seq 0 2`
+    do
+        line=`cat $CONFD_XC | grep "${srvc_ids[$i]}" | grep "r"`
+        M0D_SERV_FIDS[$i]=`echo $line | awk '{ print $2; }' | awk '{ print substr($1, 3, length($1) - 4); }'`
+        M0D_ENDPOINTS[$i]=`echo $line | awk '{ print $12; }' | awk -F":" '{ printf("%s:%s:%s", $2, $3, substr($4, 0, length($4) - 2));}'`
+    done
+}
+
+function ha_msg_send_transient()
```
Comment (just-now) [inline dtm0/it/all2all/all2all:154]: use loop.

File: dtm0/it/all2all/all2all [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+{
+    $MOTR_ST_UTILS_DIR/ha_msg_send.sh "${M0D_ENDPOINTS[0]}" "${M0D_SERV_FIDS[1]}" "transient"
+    $MOTR_ST_UTILS_DIR/ha_msg_send.sh "${M0D_ENDPOINTS[1]}" "${M0D_SERV_FIDS[0]}" "transient"
+    $MOTR_ST_UTILS_DIR/ha_msg_send.sh "${M0D_ENDPOINTS[2]}" "${M0D_SERV_FIDS[0]}" "transient"
+}
+
+function ha_msg_send_online()
```
Comment (just-now) [inline dtm0/it/all2all/all2all:161]: loop

File: dtm0/it/all2all/all2all [hunk 20]
```diff
@@ -0,0 +1,239 @@
...
+    $MOTR_ST_UTILS_DIR/ha_msg_send.sh "${M0D_ENDPOINTS[0]}" "${M0D_SERV_FIDS[2]}" "online"
+
+    $MOTR_ST_UTILS_DIR/ha_msg_send.sh "${M0D_ENDPOINTS[1]}" "${M0D_SERV_FIDS[0]}" "online"
+    $MOTR_ST_UTILS_DIR/ha_msg_send.sh "${M0D_ENDPOINTS[1]}" "${M0D_SERV_FIDS[2]}" "online"
+
+    $MOTR_ST_UTILS_DIR/ha_msg_send.sh "${M0D_ENDPOINTS[2]}" "${M0D_SERV_FIDS[0]}" "online"
+    $MOTR_ST_UTILS_DIR/ha_msg_send.sh "${M0D_ENDPOINTS[2]}" "${M0D_SERV_FIDS[1]}" "online"
+}
+
+function check_traces_ne()
```
Comment (just-now) [inline dtm0/it/all2all/all2all:173]: name should be different.

**Pull request merged**